healthcheck/check.go

```diff
@@ -24,36 +44,178 @@ type CheckState struct {
 
 // Check represents a check performed by the health check
 type Check struct {
-	State   *CheckState
-	Checker *Checker
-	mutex   *sync.Mutex
+	state   *CheckState
+	checker Checker
+}
+
+// Name gets the check name
+func (s *CheckState) Name() string {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	return s.name
+}
+
+// Status gets the check status
+func (s *CheckState) Status() string {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	return s.status
+}
+
+// StatusCode gets the check status code
+func (s *CheckState) StatusCode() int {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	return s.statusCode
+}
+
+// Message gets the check message
+func (s *CheckState) Message() string {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	return s.message
+}
+
+// LastChecked gets the last checked time of the check
+func (s *CheckState) LastChecked() *time.Time {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	if s.lastChecked == nil {
+		return nil
+	}
+
+	t := *s.lastChecked
+	return &t
+}
+
+// LastSuccess gets the time of the last successful check
+func (s *CheckState) LastSuccess() *time.Time {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	if s.lastSuccess == nil {
+		return nil
+	}
+
+	t := *s.lastSuccess
+	return &t
+}
+
+// LastFailure gets the time of the last failed check
+func (s *CheckState) LastFailure() *time.Time {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	if s.lastFailure == nil {
+		return nil
+	}
+
+	t := *s.lastFailure
+	return &t
+}
+
+// Update updates the relavent state fields based on the status provided
+// name of the check
+// status of the check, must be one of healthcheck.StatusOK, healthcheck.StatusWarning or healthcheck.StatusCritical
+// message briefly describing the check state
+// statusCode returned if the check was an HTTP check (optional, provide 0 if not relevant)
+func (s *CheckState) Update(name, status, message string, statusCode int) error {
+	now := time.Now().UTC()
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	switch status {
+	case StatusOK:
+		s.lastSuccess = &now
+	case StatusWarning, StatusCritical:
+		s.lastFailure = &now
+	default:
+		return fmt.Errorf("invalid check status, must be one of %s, %s or %s", StatusOK, StatusWarning, StatusCritical)
+	}
+
+	s.status = status
+	s.message = message
+	s.statusCode = statusCode
+	s.lastChecked = &now
+
+	if s.name == "" {
+		s.name = name
+	}
+
+	return nil
+}
+
+// hasRun returns true if the check has been run and has state
+func (c *Check) hasRun() bool {
+	if c.state.LastChecked() == nil {
+		return false
+	}
+	return true
 }
 
 // newCheck returns a pointer to a new instantiated Check with
 // the provided checker function
-func newCheck(checker *Checker) (*Check, error) {
+func newCheck(checker Checker) (*Check, error) {
```

Comment:
Can/should this be exported?

Reply:
It could be.  I don't think there is any harm so might as well 👍

Reply:
So I can't see a way to create the mutex on a CheckState object.

I would need this to be able to instantiate a valid (yet empty) check object, mutex automatically gets set to nil and I don't want this as I want to call the getters and setters against checkState object

Reply:
Are you trying to create a new `CheckState` or `Check` object?

Reply:
An initialised `CheckState` object to pass into Checker function